app/ide-desktop/lib/dashboard/src/services/LocalBackend.ts

```diff
@@ -271,25 +457,35 @@ export default class LocalBackend extends Backend {
    * @throws An error if the JSON-RPC call fails. */
   override async deleteAsset(
     assetId: backend.AssetId,
-    _force: boolean,
+    body: backend.DeleteAssetRequestBody,
     title: string | null
   ): Promise<void> {
-    // This is SAFE, as the only asset type on the local backend is projects.
-    // eslint-disable-next-line no-restricted-syntax
-    const projectId = assetId as backend.ProjectId
-    if (LocalBackend.currentlyOpeningProjectId === projectId) {
-      LocalBackend.currentlyOpeningProjectId = null
-    }
-    LocalBackend.currentlyOpenProjects.delete(projectId)
-    try {
-      await this.projectManager.deleteProject({ projectId })
-      return
-    } catch (error) {
-      throw new Error(
-        `Could not delete project ${title != null ? `'${title}'` : `with ID '${projectId}'`}: ${
-          errorModule.tryGetMessage(error) ?? 'unknown error'
-        }.`
-      )
+    const typeAndId = extractTypeAndId(assetId)
+    switch (typeAndId.type) {
+      case backend.AssetType.directory:
+      case backend.AssetType.file: {
+        await this.runProjectManagerCommand(null, 'filesystem-delete', typeAndId.id)
+        return
+      }
+      case backend.AssetType.project: {
+        if (LocalBackend.currentlyOpeningProjectId === typeAndId.id) {
+          LocalBackend.currentlyOpeningProjectId = null
+        }
+        LocalBackend.currentlyOpenProjects.delete(newProjectId(typeAndId.id))
+        try {
+          await this.projectManager.deleteProject({
+            projectId: typeAndId.id,
+            projectsDirectory: extractTypeAndId(body.parentId).id,
+          })
+          return
+        } catch (error) {
+          throw new Error(
+            `Could not delete project ${
+              title != null ? `'${title}'` : `with ID '${typeAndId.id}'`
+            }: ${errorModule.tryGetMessage(error) ?? 'unknown error'}.`
+          )
+        }
+      }
```

Comment:
This is what I mean by two sources of truth.

If we had a `LocalBackend.projects` then `LocalBackend.currentlyOpeningProjectId` could be materialized by filtering over `LocalBackend.projects` and we wouldn't have to update it separately from `LocalBackend.currentlyOpenProjects`.

Setting `project.projectState = closed` on `LocalBackend.projects` would be sufficient to update both properties in that case.